### ISSUE TITLE
Issues remain for handling development releases

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/chartpress#177]
### Background

`chartpress` has relied on information in git tags to deduce development chart versions and image tags, but we realized a need to guide chartpress more explicitly and provide it a _base version_ of some kind. In #150 we introduced the `useChartVersion` option that made Chart.yaml's version field be the source of the _base version_.

Keeping state in Chart.yaml's version field is a bit complicated though.
- Having Chart.yaml's version field as the source of the base version, but not exactly the base version, makes us need to handle logic to deduce the base version.
- With #150 we made `chartpress --reset` more complicated and influenced by useChartVersion` being true or false.
- With #150 we end with a workflow that pushes default versions and values that are not published. For example, we would have `2.0.0-0.dev` as a version in z2jh's Chart.yaml's version field (and image tags), but they are not available. It could be less confusing if it instead said something like `set-by-chartpress` to indicate a need to use the tool chartpress.

### Alternative proposal (1)

- We start speaking about a _base version_, which typically chartpress has calculated as the latest git tag on branch or `0.0.1` as a fallback.
  - Benefit: easy to understand this concept
- We let the _base version_ be possible to explicitly override in `chartpress.yaml` under `<chart>.baseVersion`.
  - Benefit: I think it will be relatively easy to understand this configurations impact, as well as implementing it and maintaining it
- We require `<chart>.baseVersion` to be a pre-release if specified, and document examples like `2.0.0-0.dev` or `2.0.0-beta.1` 
  - Benefit: we end up not needing logic to add `-0.dev` in chartpress itself
  - Benefit: we end up being able to speak about a _chartpress suffix_ or _chartpress development release suffix_ to mean specifically the `.git.n.hash` part and not add confusion about it also meaning `-0.dev` or similar.

#### Proposed workflow with development releases

- Initially set `baseVersion` in `chartpress.yaml` to your planned next release.

  ```yaml
  charts:
    - name: mychart
      baseVersion: 1.2.0-0.dev
  ```

- When a release is to be made associated with a git tag, just push a git tag and trigger automation configured to run `chartpress --push --publish-chart --tag <tag>`

  ```bash
  V=1.2.0
  git commit -am "release $V"
  git tag -am "release $V" "$V"
  git push --atomic --follow-tags
  ```

- After a major or minor release (such as 1.2.0) or because of a change of plans, update the configured base version to reflect the next minor version you plan to release (such as 1.3.0, or 2.0.0).

  ```bash
  # update chartpress.yaml's <chart>.baseVersion to 1.3.0-0.dev
  git commit -am "Declare next major.minor version planned"
  git push --atomic
  ```

#### Proposed action points

- Revert `chartpress --reset --tag <tag>` functionality and rely on `chartpress --skip-build --tag <tag>` for the same functionality, as tracked in #176.
- Revert `useChartVersion` functionality and implement alternative proposal

### Insights into problems (gained after writing proposal above)

I've sadly realized two problems #150 and this alternative proposal will fail to address.

- __Implicit `--tag`__ (not so severe)
  I assume users may end up with a `baseVersion` older than the latest available git tag, for example in automation if users run `chartpress --push --publish-chart` without specifying `--tag <git tag>` explicitly.
- __Reset n-counts__ (severe)
  I assume users may publish unplanned patch releases with git tags directly from the default git branch, if they do, we have big trouble.
  - Example: consider 1.2.0 is released and baseVersion is declared to be 1.3.0-0.dev, then we will get development release versions like `1.3.0-0.dev.git.5.h123` for the fifth commit since 1.2.0, but that will also be the name for the fifth commit since 1.2.1 later released.

#### Ideas on handling reset n-counts problem

I think the purpose of the `n` field in the chartpress suffix is to ensure that we have something that is reliably incrementing for a given base version, so that development releases are sorted reliably internally. Practically I think if we would count the commits since a baseVersion was first used, we would be fine.

_Brainstorm_
- We could ask users to manually declare a previous git tag or git reference to compare against
  - issue: adds complexity to the UX
  - benefit: it would be reliable
  - possible issue: we may not have a good way to get `n` from something besides the latest commit, maybe we do, I'm not sure
- We could git blame the line where baseVersion was set to get a commit and count from that
  - issue: if someone added quotes to the YAML file line where baseVersion was specified, or if someone declared the yaml on a single line of code etc, we would be in trouble.
  - issue: detecting a line where baseVersion is within a YAML file would be messy and prone to issues
- We could try deduce a tag to compare with based on the baseVersion tag.
  - issue: not able to come up with an algorithm to deduce a tag to calculate an n-count from that would lead to a strictly incrementing number
- We could calculate n-count from the initial commit
  - non critical issue: the n-count is less convenient as an indicator of relative progress since last release

_Conclusion_

I'm not a fan of any solution, but appreciate the strategy of counting from the initial commit on branch the most among the options assuming we have an explicitly specified `baseVersion` to relate to.

### Final notes for now

I have no clear path forward here, at the same time I feel that we can't release 2.0.0 and promote use of `useChartVersion` as it has the _reset n-count_ problem.

**Comment by minrk** [comment]
Thanks for thinking this through so carefully!

I don't quite follow the implicit `--tag` issue. Can you describe it more?


I think it you're right that it will simplify things by moving the baseVersion to chartpress.yaml instead of using things from Chart.yaml. We can also make it clear that we only ever consider this version when publishing development releases. It's also completely irrelevant when only publishing tags.

If we put both `baseVersion` and `previousTag` in chartpress.yaml, I think we could get to what you are after in your conclusion. I think the `previousTag` part would be _more_ complex for both chartpress users and maintainers, though not intractable. I'm not sure the trade-off is worth it to solve the n-count reset issue, but we can look into it.

I agree that the n-count issue results in imperfect counters when a less-than-next patch release is made from a branch, but disagree that this is a serious issue. It can only, by definition, happen soon after the previous release. So the overlaps cannot ever be large, and the full names cannot collide because they include the commit hash. Plus, versions affected are necessarily always development releases, which should be considered stale almost immediately on publication of the next version anyway. Even with that, the situation seems to me significantly improved over the current behavior. We already have the same n-commit count overlap issue for backport branches because two branches are counting from the same tag and it's never come up, and worse because once there's a backport release commits on the backport branch will be published with greater _version_ tags than main, not just commit counters, which are only ever meant as a hint.

**Comment by minrk** [comment]
I started looking at moving baseVersion to chartpress.yaml and it is indeed way simpler. Lots of checks and branches just go away. Great idea! I'll have a PR tomorrow.

**Comment by consideRatio** [comment]
## About _reset n-count_ issue

> I agree that the n-count issue results in imperfect counters when a less-than-next patch release is made from a branch, but disagree that this is a serious issue. It can only, by definition, happen soon after the previous release.

I don't think I understand this statement clearly, hmmm... Let me clarify the problematic situation I imagine:

1. There is no backport branches etc, just the main branch
2. 1.0.0 is released, and a baseVersion is declared to be `1.1.0-0.dev`
3. Small PRs not motivating a minor version bump are merged, and we get development releases like for example `1.1.0-0.dev.git.10.h123` and `1.1.0-0.dev.git.20.h456`.
4. A critical bug is patched and we want to go for 1.0.1 straight from the main branch.
5. Small PRs not motivating a minor version bump are merged, and we get development releases like for example `1.1.0-0.dev.git.10.habc` and `1.1.0-0.dev.git.20.hdef`.

This can keep happening, and when 1.0.2 is released, we will have `1.1.0-0.dev` releases with n-counts from the `1.0.0` tag, the `1.0.1` tag, and the `1.0.2` tag, all overlapping and looking like `1.1.0-0.dev.git.n.hash`. That way, we can't say if for example `1.1.0-0.dev.git.10.habc` is newer than `1.1.0-0.dev.git.20.h456` without comparing dates associated with the git commit.

### Suggested resolution of the _reset n-count_ issue: _full n-count_

I suggest we stop counting commits since the latest tag on a branch as that leads to the _reset n-count_ issue, and we instead count all the commits since the initial commit on the branch (`git rev-list --count HEAD`) and use that as our n-count.

## About _Implicit --tag_ issue

> I don't quite follow the implicit --tag issue. Can you describe it more?

I'm unsure if this will happen or if the consequences are problematic or not, but the issue is that in the past the chartpress suffix was oimtted by default if a commit was tagged, even though `--tag` wasn't specified. Now though with an explicit base version, omitting `--tag` would mean we go for a chartpress suffix related to the base version. 

So, I think any release automation that didn't explicitly declare `--tag` when a tag was pushed could fail to package a version as clean as the git tag.

To conclude, historically omitting `--tag <git tag>` was a way to version a chart like the git tag, but now it won't be with a baseVersion unless we check for that. It may not be a big problem, but it is change that may be unexpected. z2jh, binderhub, dask/dask-gateway all specify `--tag` explicitly when running for a tagged git commit though.

**Comment by minrk** [comment]
> Now though with an explicit base version, omitting --tag would mean we go for a chartpress suffix related to the base version.

Ah, I follow now. I think we should preserve the "git tag == chartpress tag" behavior. I'll make sure we have test coverage for that.

> we instead count all the commits since the initial commit on the branch (git rev-list --count HEAD) and use that as our n-count.

I think that's a terrific idea - solves the reset _and_ doesn't introduce the complexity of having to count commits ourselves / requiring users to specify where to count from. I didn't know about `rev-list --count`!

**Comment by consideRatio** [comment]
I think this is all resolved by #178 and #179! Thank you @minrk for resolving this so thoroughly!

Should we go for a release of 2.0.0? I can write changelog etc

**Comment by minrk** [comment]
> Should we go for a release of 2.0.0

I believe so

>  I can write changelog etc

That would be great, thank you!
